refactor(transformer/class-properties): re-order fields and expand comments

overlookmotel opened the pull request:
Follow-on after #10418. Pure refactor.

`private_field_count` is accessed in both the entry and exit phases of the transform, so move that field up to be with the other fields which are in that same category.

Expand the comment which explains how this optimization works.

Also, when reviewing #10418, I had to trace through the logic to make sure that it wasn't possible for decreasing `private_field_count` on exit to get skipped if `is_transform_required` is `false`. Turns out it's not a problem - `is_transform_required` is always `true` if class has any private properties/methods. Add comments to clarify that.

overlookmotel commented:
* **#10447** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10447?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10447?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=10447).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Dunqing reviewed (approved):
Looks good!

By the way, I forgot one thing in my PR: to add a `debug_assert_eq!(self.private_field_count, 0)` in `exit_program` or elsewhere to ensure we are correctly handling `private_field_count` so that we can catch any potential bug related to this in the future if we do a huge refactor in class_properties.

overlookmotel commented:
### Merge activity

* **Apr 16, 12:27 PM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
refactor(transformer/class-properties): re-order fields and expand comments (#10447)

Follow-on after #10418. Pure refactor.

`private_field_count` is accessed in both the entry and exit phases of the transform, so move that field up to be with the other fields which are in that same category.

Expand the comment which explains how this optimization works.

Also, when reviewing #10418, I had to trace through the logic to make sure that it wasn't possible for decreasing `private_field_count` on exit to get skipped if `is_transform_required` is `false`. Turns out it's not a problem - `is_transform_required` is always `true` if class has any private properties/methods. Add comments to clarify that.
